### PR TITLE
Fix clock thread issue (#1266)

### LINKED ISSUE #1266: SIGSEGV at set_clock
## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 20.04
- Installation type:
  - Foxy release src build
- Version or commit hash:
  - rclcpp 2.0.2
- DDS implementation:
  - Fast-DDS v2.0.1
- Client library (if applicable):
  - Navigation2

#### Steps to reproduce issue

It is hard to reproduce the problem intentionally but happens in some nodes randomly.
Here is the backtrace.

```
#0  0x0000000000000031 in ?? ()
#1  0x00007efd1b6ae1f0 in rcl_clock_call_callbacks () from /opt/ros/foxy/rcl/lib/librcl.so
#2  0x00007efd1b6ae5ef in rcl_set_ros_time_override () from /opt/ros/foxy/rcl/lib/librcl.so
#3  0x00007efd1be8a729 in rclcpp::TimeSource::set_clock(std::shared_ptr<builtin_interfaces::msg::Time_<std::allocator<void> > >, bool, std::shared_ptr<rclcpp::Clock>) () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#4  0x00007efd1be8a96d in rclcpp::TimeSource::clock_cb(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >) () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#5  0x00007efd1bea591e in void std::__invoke_impl<void, void (rclcpp::TimeSource::*&)(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >), rclcpp::TimeSource*&, std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > > >(std::__invoke_memfun_deref, void (rclcpp::TimeSource::*&)(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >), rclcpp::TimeSource*&, std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >&&) () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#6  0x00007efd1bea3a0a in std::__invoke_result<void (rclcpp::TimeSource::*&)(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >), rclcpp::TimeSource*&, std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > > >::type std::__invoke<void (rclcpp::TimeSource::*&)(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >), rclcpp::TimeSource*&, std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > > >(void (rclcpp::TimeSource::*&)(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >), rclcpp::TimeSource*&, std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >&&) () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#7  0x00007efd1bea1a64 in void std::_Bind<void (rclcpp::TimeSource::*(rclcpp::TimeSource*, std::_Placeholder<1>))(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >)>::__call<void, std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >&&, 0ul, 1ul>(std::tuple<std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >&&>&&, std::_Index_tuple<0ul, 1ul>) ()
   from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#8  0x00007efd1be9fbae in void std::_Bind<void (rclcpp::TimeSource::*(rclcpp::TimeSource*, std::_Placeholder<1>))(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >)>::operator()<std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >, void>(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >&&) () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#9  0x00007efd1be9da89 in std::_Function_handler<void (std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >), std::_Bind<void (rclcpp::TimeSource::*(rclcpp::TimeSource*, std::_Placeholder<1>))(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >)> >::_M_invoke(std::_Any_data const&, std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >&&) ()
   from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#10 0x00007efd1bed378b in std::function<void (std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >)>::operator()(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >) const () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#11 0x00007efd1bed0ea7 in rclcpp::AnySubscriptionCallback<rosgraph_msgs::msg::Clock_<std::allocator<void> >, std::allocator<void> >::dispatch(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >, rclcpp::MessageInfo const&) () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#12 0x00007efd1becdf02 in rclcpp::Subscription<rosgraph_msgs::msg::Clock_<std::allocator<void> >, std::allocator<void>, rclcpp::message_memory_strategy::MessageMemoryStrategy<rosgraph_msgs::msg::Clock_<std::allocator<void> >, std::allocator<void> > >::handle_message(std::shared_ptr<void>&, rclcpp::MessageInfo const&) () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#13 0x00007efd1bd78b46 in rclcpp::Executor::execute_subscription(std::shared_ptr<rclcpp::SubscriptionBase>)::{lambda()#6}::operator()() const () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#14 0x00007efd1bd7b9bd in std::_Function_handler<void (), rclcpp::Executor::execute_subscription(std::shared_ptr<rclcpp::SubscriptionBase>)::{lambda()#6}>::_M_invoke(std::_Any_data const&) () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#15 0x00007efd1bd5c1b4 in std::function<void ()>::operator()() const () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#16 0x00007efd1bd781f8 in take_and_do_error_handling(char const*, char const*, std::function<bool ()>, std::function<void ()>) () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#17 0x00007efd1bd79216 in rclcpp::Executor::execute_subscription(std::shared_ptr<rclcpp::SubscriptionBase>) () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#18 0x00007efd1bd77f45 in rclcpp::Executor::execute_any_executable(rclcpp::AnyExecutable&) () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#19 0x00007efd1bd8595c in rclcpp::executors::SingleThreadedExecutor::spin() () from /opt/ros/foxy/rclcpp/lib/librclcpp.so
#20 0x00007efd1b512b4d in std::thread::_State_impl<std::thread::_Invoker<std::tuple<nav2_util::NodeThread::NodeThread(std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>)::{lambda()#1}> > >::_M_run() () from /opt/overlay_ws/install/nav2_util/lib/libnav2_util_core.so
#21 0x00007efd1b217cb4 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#22 0x00007efd1b32b609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#23 0x00007efd1af05103 in clone () from /usr/lib/x86_64-linux-gnu/libc.so.6
```

#### bug fix

I think the clock needs to be locked before calling rcl_set_ros_time_override which eventually access jump_callbacks array.
I have executed my simulation overnight with the fixed code and so far it looks good. 
https://github.com/daisukes/rclcpp/commit/4a2579f27fe75816c9c13c818861ff0bb41b49f7

Also, I think these two lines should have a lock because it could call the same function.
https://github.com/ros2/rclcpp/blob/bbf2c4cbfb006245c784f39ee8f8f9e56431c91e/rclcpp/src/rclcpp/time_source.cpp#L278
https://github.com/ros2/rclcpp/blob/bbf2c4cbfb006245c784f39ee8f8f9e56431c91e/rclcpp/src/rclcpp/time_source.cpp#L287

### PR DESCRIPTION
**Pull request by daisukes**
Fixes #1266

### CODE REVIEW

File: rclcpp/src/rclcpp/time_source.cpp [hunk 1]
```diff
@@ -192,6 +192,7 @@ void TimeSource::set_clock(
     enable_ros_time(clock);
   }
 
+  std::lock_guard<std::mutex> clock_guard(clock->get_clock_mutex());
```
Comment (clalancette) [inline rclcpp/src/rclcpp/time_source.cpp:195]: One problem here is that the clock could, in theory, be changed between the call to `{enable,disable}_ros_time` above, and `rcl_set_ros_time_override`.

What I'll suggest is something a little different.  Create two new `private` methods to this class called `disable_ros_time_nolock()` and `enable_ros_time_nolock()`, which has the current contents of the respective methods.  Then in `disable_ros_time()`, take the lock and call `disable_ros_time_nolock()`; do the same for `enable_ros_time()`.  Finally, change `set_clock()` to take the lock right at the beginning, and then call `disable_ros_time_nolock()` or `enable_ros_time_nolock()` as appropriate.

Comment (daisukes) [reply rclcpp/src/rclcpp/time_source.cpp]: `disable_ros_time(clock)` and `enable_ros_time(clock)` are only used in `set_clock`.
So, how about merging them into `set_clock` to make things clear.

File: rclcpp/src/rclcpp/time_source.cpp [hunk 2]
```diff
@@ -185,14 +185,21 @@ void TimeSource::set_clock(
...
   if (!set_ros_time_enabled && clock->ros_time_is_active()) {
-    disable_ros_time(clock);
+    ret = rcl_disable_ros_time_override(clock->get_clock_handle());
   } else if (set_ros_time_enabled && !clock->ros_time_is_active()) {
-    enable_ros_time(clock);
+    ret = rcl_enable_ros_time_override(clock->get_clock_handle());
+  }
+  if (ret != RCL_RET_OK) {
+    rclcpp::exceptions::throw_from_rcl_error(
+      ret, "Failed to enable ros_time_override_status");
```
Comment (clalancette) [inline rclcpp/src/rclcpp/time_source.cpp:199]: Nit: In the case we disabled it, it should be:

`Failed to disable ros_time_override_status`

(this was wrong in the previous code too)

Comment (daisukes) [reply rclcpp/src/rclcpp/time_source.cpp]: fixed!

Comment (clalancette) [reply rclcpp/src/rclcpp/time_source.cpp]: Looks good, thanks.

**Comment by clalancette** [comment]
@hidmic These changes look good to me, but I'd like a double-check from you before I run CI and merge.

**Comment by clalancette** [comment]
Thanks everyone.  Here's CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11905/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6949/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9700/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11838/)

**Comment by clalancette** [comment]
All right, CI is looking good.  One thing I will point out; when building Windows with `CMAKE_BUILD_TYPE` set to None, we would get one flaky test (rclcpp_gtest_multithreaded__rmw_connext_cpp).  But we also see those flakes in https://ci.ros2.org/view/nightly/job/nightly_win_deb/1713/#showFailuresLink .  Meanwhile in the CI run above, with `CMAKE_BUILD_TYPE` set to Release, we don't see the flake.

All of that is to say is that I don't think this PR is the cause of that failing test.  So I'm going to go ahead and merge this one.  Thanks @daisukes for the fix.

**Comment by KavenYau** [comment]
@fujitatomoya @clalancette Would this be backported to `foxy`?

**Comment by clalancette** [comment]
> @fujitatomoya @clalancette Would this be backported to `foxy`?

This could be considered for a backport to Foxy.  If you'd like to see it there, I'd encourage you to open a PR.  One thing we'll have to be careful of is that this PR removes two static methods.  I'm honestly not sure if that changes API or ABI, but if it does, we'll need another solution.

**Comment by fujitatomoya** [comment]
i think that static private members can be removed w/o changing ABI (if these are not used in inline functions). so i think we can backport this to foxy.

https://github.com/ros2/rclcpp/blob/1c92622516d676a7a5f8075b241410d8d498dd9f/rclcpp/include/rclcpp/time_source.hpp#L150-L152

**Comment by fujitatomoya** [comment]
I just opened backport PR to foxy, https://github.com/ros2/rclcpp/pull/1685

**Pull request merged**